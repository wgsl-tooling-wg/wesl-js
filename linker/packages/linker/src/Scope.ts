import { DeclarationElem } from "./AbstractElems2.ts";

export interface SrcModule {
  /** module path "rand_pkg::sub::foo", or "package::main" */
  modulePath: string;
  
  /** file path to the module for user error reporting e.g "rand_pkg:sub/foo.wesl", or "./sub/foo.wesl" */
  filePath: string;

  /** original src for module */
  src: string;
}

export interface SrcLoc {
  pos: number;
  src: SrcModule;
}

/** a src declaration or reference to an ident */
export type Ident = DeclIdent | RefIdent;

export type Conditions = Record<string, boolean>;

interface IdentBase {
  originalName: string; // name in the source code for ident matching (may be mangled in the output)
  conditions?: Conditions; // conditions under which this ident is valid (combined from all containing elems)
}

export interface RefIdent extends IdentBase {
  kind: "ref";
  refersTo?: Ident; // import or decl ident in scope to which this ident refers. undefined before binding
  std?: true; // true if this is a standard wgsl identifier (like sin, or u32)
}

export interface DeclIdent extends IdentBase {
  kind: "decl";
  mangledName?: string; // name in the output code
  declElem: DeclarationElem;
}

export type ScopeKind =
  | "module" // root scope for a module (file)
  | "body"; // a scope inside the module (fn body, nested block, etc.)

/** tree of ident references, organized by lexical scope */
export interface Scope {
  id?: number; // for debugging
  idents: Ident[]; // idents found in lexical order in this scope
  parent: Scope | null; // null for root scope in a module
  children: Scope[];
  kind: ScopeKind;
}

export function resetScopeIds() { // for debugging
  scopeId = 0;
}

let scopeId = 0; // for debugging

/** make a new Scope object */
export function makeScope(s: Omit<Scope, "id">): Scope {
  return { ...s, id: scopeId++ };
}

export interface RootAndScope {
  rootScope: Scope;
  scope: Scope;
}

export function emptyScope(kind: ScopeKind): Scope {
  return makeScope({ idents: [], parent: null, children: [], kind });
}

export function emptyBodyScope(parent: Scope): Scope {
  return makeScope({ kind: "body", idents: [], parent, children: [] });
}

/** For debugging,
 * @return true if a scope is in the rootScope tree somewhere */
export function containsScope(rootScope: Scope, scope: Scope): boolean {
  if (scope === rootScope) {
    return true;
  }
  for (const child of rootScope.children) {
    if (containsScope(child, scope)) {
      return true;
    }
  }
  return false;
}

export function exportDecl(scope:Scope, name: string): DeclIdent | undefined {
  for (const ident of scope.idents) {
    if (ident.originalName === name && ident.kind === "decl") {
      return ident;      
    }
  }
}