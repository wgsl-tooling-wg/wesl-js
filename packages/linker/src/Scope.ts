import { pretty } from "berry-pretty";
import { tracing } from "mini-parse";

type IdentKind = "decl" | "ref";

export interface SrcModule {
  /** full path to the module e.g "package/sub/foo", or "_root/sub/foo" */
  modulePath: string;

  /** original src for module */
  src: string;
}

export interface SrcLoc {
  pos: number;
  src: SrcModule;
}

/** a src declaration or  */
export interface Ident {
  kind: IdentKind;
  refersTo?: Ident | null; // preceding ident in scope. null for decls, undefined before binding
  originalName: string; // name in the source code for ident matching (may be mangled in the output)
}

export type ScopeKind =
  | "module" // root scope for a module (file)
  | "body"; // a scope inside the module (fn body, nested block, etc.)

/** tree of ident references, organized by lexical scope */
export interface Scope {
  idents: Ident[]; // idents found in lexical order in this scope
  parent: Scope | null; // null for root scope in a module
  children: Scope[];
  kind: ScopeKind;
}

export interface RootAndScope {
  rootScope: Scope;
  scope: Scope;
}

/** @return a new root scope and new current scope with an ident added to the current scope. */
export function withAddedIdent(
  rootScope: Scope,
  scope: Scope,
  ident: Ident,
): RootAndScope {
  if (tracing && !containsScope(rootScope, scope)) {
    logScope("scope not in rootScope", rootScope);
    logScope("..scope", rootScope);
  }

  // clone the current provisional scope with new ident added
  const scopeIdents = scope.idents;
  const idents: Ident[] = [...scopeIdents, ident];
  const newScope: Scope = { ...scope, idents };
  const newRootScope = cloneScopeReplace(rootScope, scope, newScope);

  return {
    scope: newScope,
    rootScope: newRootScope,
  };
}

/** @return a new root scope and new current scope with a child scope added to the current scope. */
export function withChildScope(
  rootScope: Scope,
  currentScope: Scope,
  newChildScope: Scope,
): RootAndScope {
  const newCurrentScope = {
    ...currentScope,
    children: [...currentScope.children, newChildScope],
  };
  const newRootScope = cloneScopeReplace(
    rootScope,
    currentScope,
    newCurrentScope,
  );

  return {
    scope: newChildScope,
    rootScope: newRootScope,
  };
}

/** For debugging,
 * @return true if a scope is in the rootScope tree somewhere */
function containsScope(rootScope: Scope, scope: Scope): boolean {
  if (scope === rootScope) {
    return true;
  }
  for (const child of rootScope.children) {
    if (containsScope(child, scope)) {
      return true;
    }
  }
  return false;
}

/** clone the rootScope, replacing oldScope with newScope in child and parent links */
function cloneScopeReplace(
  rootScope: Scope,
  oldScope: Scope,
  newScope: Scope,
): Scope {
  if (rootScope === oldScope) {
    return newScope;
  }

  const { kind, idents } = rootScope;
  const { parent: rootParent, children: rootChildren } = rootScope;
  const parent = rootParent === oldScope ? newScope : rootParent;
  const children = rootChildren.map(s =>
    cloneScopeReplace(s, oldScope, newScope),
  );
  return { kind, idents, parent, children };
}

export function logScope(message: string, scope: Scope) {
  console.log(`${message}:`);
  console.log(scopeToString(scope, 2));
}

export function scopeToString(scope: Scope, indent = 0): string {
  const { kind, idents, children } = scope;
  const identStr = pretty(idents.map(i => i.originalName));
  const childStrings = children.map(c => scopeToString(c, indent + 4));
  const childrenStr = childStrings.join("\n");
  const spc = " ".repeat(indent);
  // prettier-ignore
  return `${spc}${kind} ${identStr}\n` + 
         `${spc}  children:\n`+ 
         `${childrenStr}`;
}

// export interface ProvisionalScope {
//   currentScope: Scope; // current scope at this level of parsing
//   idents: Ident[]; // idents pending to add to current scope
//   scopes: Scope[]; // scopes pending to add to current scope

// }

// TODO do we need to bother with the tricky indexing to symbols?
// . two level table seems relevant for multithreading..
//   . also useful for say watch mode
//   . lets do it
// . and why do we need to use indices into the table vs pointers?
//   . if we want to replace a symbol we could just mutate it in place
//   . (esbuild claims its handy to clone the symbol table, so index references stay valid)
//   . lets skip it til we know we need it
// . and why do we need a single global symbol table to collect all the symbols
//   . scope tables seem sufficient..
//   . lets skip it til we know we need it
//
// let's build some stuff and find out..

/* 
Key tasks:
- accumulate idents and scopes during the parse
- bind idents references to declarations
- traverse AST from root module to emit
  - store idents in the AST? and references to AST from idents?
  - That would let us traverse the AST to find referenced AST elements
    - e.g. fn foo() { bar();} we need to traverse to bar()

Datastructures should be compatible for future parallelization:
- parse files in parallel
  - each file gets its own ident table
- presume ident binding is single threaded for now
  - esbuild is single threaded (according to arch diagram)
  - do any necessary ident mangling
  - works entirely on ident table
- emit code in parallel
  - 

*/

// export interface AllIdents {
//   tables: IdentTable[];   // idents for all tables
// }

// export interface IdentTable {
//   idents: Ident[][]; // indexed by moduleIndex and identIndex
// }

// // reference to a symbol in the Ident table, e.g. from an AST
// export interface IdentRef {
//   moduleIndex: number; // index of the file
//   identIndex: number; // index of the identifier within the file
// }
