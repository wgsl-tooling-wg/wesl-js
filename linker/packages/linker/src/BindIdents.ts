import { debugNames } from "mini-parse";
import { DeclarationElem } from "./AbstractElems2.ts";
import { FlatImport } from "./FlattenTreeImport.ts";
import { ParsedRegistry2 } from "./ParsedRegistry2.ts";
import { DeclIdent, exportDecl, RefIdent, Scope } from "./Scope.ts";
import { identToString, scopeIdentTree } from "./ScopeLogging.ts";
import { stdFn, stdType } from "./TraverseRefs.ts";
import { last, overlapTail } from "./Util.ts";
import { dlog } from "berry-pretty";
import { flatImports, WeslAST } from "./ParseWESL.ts";

/**
 * Bind active reference idents to declaration Idents by mutating the refersTo: field
 * Also in this pass, set the mangledName: field for all active global declaration idents.
 *
 * @param parsed
 * @param conditions  only bind to/from idents that are valid with the current condition set
 * @return any new declaration elements found (they will need to be emitted)
 */
export function bindIdents(
  ast: WeslAST,
  parsed: ParsedRegistry2,
  conditions: Record<string, any>,
): DeclarationElem[] {
  /* 
    For each module's scope, search through the scope tree to find all ref idents
      - For each ref ident, search up the scope tree to find a matching decl ident
      - If no local match is found, check for partial matches with import statements
        - combine ident with import statement to match a decl in exporting module

    As global decl idents are found, mutate their mangled name to be globally unique.
*/
  const { rootScope } = ast;

  const globalNames = new Set<string>();
  const knownDecls = new Set<DeclIdent>();
  rootScope.idents.forEach(ident => {
    if (ident.kind === "decl") {
      ident.mangledName = ident.originalName;
      globalNames.add(ident.originalName);
      knownDecls.add(ident);
    }
  });

  const bindContext = {
    parsed,
    conditions,
    knownDecls,
    foundScopes: new Set<Scope>(),
    globalNames,
  };
  const decls = bindIdentsRecursive(rootScope, bindContext);
  return decls.flatMap(d =>
    d.declElem && isGlobal(d.declElem) ? [d.declElem] : [],
  );
}

interface BindContext {
  parsed: ParsedRegistry2;
  conditions: Record<string, any>;
  knownDecls: Set<DeclIdent>; // decl idents discovered so far
  foundScopes: Set<Scope>; // save work by not processing scopes multiple times
  globalNames: Set<string>; // root level names  used so far
}

/**
 * Recursively bind references to declarations in this scope and
 * any child scopes referenced by these declarations.
 * Uses a hash set of found declarations to avoid duplication
 * @ return any new declarations found
 */
function bindIdentsRecursive(
  scope: Scope,
  bindContext: BindContext,
): DeclIdent[] {
  // early exist if we've processed this scope before
  const { foundScopes } = bindContext;
  if (foundScopes.has(scope)) return [];
  foundScopes.add(scope);

  // dlog(scopeIdentTree(scope));

  const { parsed, conditions } = bindContext;
  const { globalNames, knownDecls } = bindContext;
  const newDecls: DeclIdent[] = []; // new decl idents to process (and return)

  scope.idents.forEach((ident, i) => {
    // dlog(`--- considering ident ${identToString(ident)}`);
    if (ident.kind === "ref") {
      if (!ident.refersTo && !ident.std) {
        if (stdWgsl(ident.originalName)) {
          ident.std = true;
        } else {
          let foundDecl =
            findDeclInModule(scope, ident, i) ?? findDeclImport(ident, parsed);

          if (foundDecl && !knownDecls.has(foundDecl)) {
            // dlog(
            //   `  > found decl: ${identToString(foundDecl)} known: ${knownDecls.has(foundDecl)}`,
            // );
            setDisplayName(ident.originalName, foundDecl, globalNames);
            // dlog(`  > queuing new decl: ${identToString(foundDecl)}`);
            if (foundDecl.declElem && isGlobal(foundDecl.declElem)) {
              newDecls.push(foundDecl);
            }
            knownDecls.add(foundDecl);
          }
          bindRefToDecl(ident, foundDecl, knownDecls);
        }
      }
    }
  });

  // follow references from child scopes
  const newFromChildren = scope.children.flatMap(child => {
    // dlog("newFromChildren", { childScope: scopeIdentTree(child) });
    return bindIdentsRecursive(child, bindContext);
  });
  // console.log(
  //   "new from children",
  //   newFromChildren.map(d => identToString(d)),
  // );

  // follow references from referenced declarations
  const newFromRefs = newDecls.flatMap(decl => {
    // dlog("newFromRefs", {
    //   decl: identToString(decl),
    //   declScope: scopeIdentTree(decl.scope),
    // });
    if (debugNames && !decl.scope) {
      console.log(`--- decl ${identToString(decl)} has no scope`);
      return [];
    }
    return bindIdentsRecursive(decl.scope, bindContext);
  });
  // console.log(
  //   "new from refs",
  //   newFromRefs.map(d => identToString(d)),
  // );

  return [newDecls, newFromChildren, newFromRefs].flat();
}

function setDisplayName(
  proposedName: string,
  decl: DeclIdent,
  globalNames: Set<string>,
): void {
  if (!decl.mangledName) {
    // if (!decl.declElem) {
    //   console.log(
    //     `--- decl ident ${identToString(decl)} has no declElem attached`,)
    // } else
    if (decl.declElem && isGlobal(decl.declElem)) {
      decl.mangledName = declUniqueName(proposedName, globalNames);
      // dlog(`  > mangle global decl: ${identToString(decl)}`);
    } else {
      // dlog(`  > no-mangle local decl: ${identToString(decl)}`);
      decl.mangledName = decl.originalName;
    }
  }
}

function bindRefToDecl(
  ident: RefIdent,
  foundDecl: DeclIdent | undefined,
  knownDecls: Set<DeclIdent>,
) {
  // dlog({ ident: identToString(ident), foundDecl: identToString(foundDecl) });
  if (foundDecl) {
    ident.refersTo = foundDecl;

    knownDecls.add(foundDecl);
  } else {
    // TODO log error with source position
    console.log(`--- unresolved ident: ${ident.originalName}`);
  }
}

function stdWgsl(name: string): boolean {
  return stdType(name) || stdFn(name);
}

/** search earlier in the scope and in parent scopes to find a matching decl ident */
function findDeclInModule(
  scope: Scope,
  ident: RefIdent,
  identDex: number = scope.idents.length,
): DeclIdent | undefined {
  const { idents, parent } = scope;
  const { originalName } = ident;

  // dlog(`  |> findDeclInModule ${originalName}`);
  // see if the declaration is in this scope
  for (let i = identDex - 1; i >= 0; i--) {
    const checkIdent = idents[i];
    const { kind } = checkIdent;
    if (kind === "decl" && originalName === checkIdent.originalName) {
      // dlog(`  |> found decl in scope: ${identToString(checkIdent)}`);
      return checkIdent;
    }
  }

  // recurse to check all idents in parent scope
  if (parent) {
    return findDeclInModule(parent, ident);
  }
}

/** Match a reference identifier to a declaration in
 * another module via an import statement */
function findDeclImport(
  refIdent: RefIdent,
  parsed: ParsedRegistry2,
): DeclIdent | undefined {
  // dlog(identToString(refIdent), { ast: !!refIdent.ast });
  const flatImps = flatImports(refIdent.ast);

  // find module path by combining identifer reference with import statement
  const modulePathParts = matchingImport(refIdent, flatImps); // module path in array form

  if (modulePathParts) {
    return findExport(modulePathParts, parsed);
  }
}

/** using the flattened import array, find an import that matches a provided identifier */
function matchingImport(
  ident: RefIdent,
  flatImports: FlatImport[],
): string[] | undefined {
  const identParts = ident.originalName.split("::");
  for (const flat of flatImports) {
    const impTail = overlapTail(flat.importPath, identParts);
    if (impTail) {
      return [...flat.modulePath, ...impTail];
    }
  }
}

/** @return an exported root element for the provided path */
function findExport(
  modulePathParts: string[],
  parsed: ParsedRegistry2,
): DeclIdent | undefined {
  const legacyConvert = modulePathParts.map(p => (p === "." ? "package" : p)); // TODO rm after we update grammar
  const modulePath = legacyConvert.slice(0, -1).join("::");
  const module = parsed.modules[modulePath];
  if (!module) {
    console.log(
      `ident ${modulePathParts.join("::")} in import statement, but module not found`,
    );
  }

  return exportDecl(module.rootScope, last(modulePathParts)!);
}

/** return mangled name for decl ident,
 *  mutating the Ident to remember mangled name if it hasn't yet been determined */
export function declUniqueName(
  proposedName: string,
  rootNames: Set<string>,
): string {
  const displayName = uniquifyName(proposedName, rootNames);
  rootNames.add(displayName);

  return displayName;
}

/** construct global unique name for use in the output */
function uniquifyName(proposedName: string, rootNames: Set<string>): string {
  let renamed = proposedName;
  let conflicts = 0;

  // create a unique name
  while (rootNames.has(renamed)) {
    renamed = proposedName + conflicts++;
  }

  // dlog({ proposedName, renamed });
  return renamed;
}

export function isGlobal(elem: DeclarationElem): boolean {
  return ["alias", "const", "override", "fn", "struct", "gvar"].includes(
    elem.kind,
  );
}
