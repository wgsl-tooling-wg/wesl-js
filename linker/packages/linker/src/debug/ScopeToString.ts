import { LineWrapper } from "./LineWrapper.ts";
import { Ident, Scope } from "../Scope.ts";

/** A debugging print of the scope tree with identifiers in nested brackets */
export function scopeToString(scope: Scope, indent = 0): string {
  const { children } = scope;
  let childStrings: string[] = [];
  if (children.length)
    childStrings = children.map(c => scopeToString(c, indent + 2));

  // list of identifiers, with decls prefixed with '%'
  const identStrings = scope.idents.map(({ kind, originalName }) => {
    const prefix = kind === "decl" ? "%" : "";
    return `${prefix}${originalName}`;
  });

  const str = new LineWrapper(indent);
  str.add("{ ");

  const last = identStrings.length - 1;
  identStrings.forEach((s, i) => {
    const element = i < last ? s + ", " : s;
    str.add(element);
  });

  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"));
  }

  if (str.oneLine) {
    str.add(" }");
  } else {
    if (!childStrings.length) str.nl();
    str.add("}");
  }

  return str.result;
}

export function identToString(ident?: Ident): string {
  if (!ident) return JSON.stringify(ident);
  const { kind, originalName } = ident;
  const idStr = ident.id ? `#${ident.id}` : "";
  if (kind === "ref") {
    const ref = identToString(ident.refersTo!);
    return `${originalName} ${idStr} -> ${ref}`;
  } else {
    return `%${originalName} ${idStr} (${ident.mangledName})`;
  }
}
