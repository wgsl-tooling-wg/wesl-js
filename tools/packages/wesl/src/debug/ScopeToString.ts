import { childScope, Ident, Scope } from "../Scope.ts";
import { attributeToString } from "./ASTtoString.ts";
import { LineWrapper } from "./LineWrapper.ts";

/** A debugging print of the scope tree with identifiers in nested brackets */
export function scopeToString(scope: Scope, indent = 0): string {
  const { contents, kind, ifAttributes = [] } = scope;

  const str = new LineWrapper(indent);
  const attrStrings = ifAttributes.map(a => attributeToString(a)).join(" ");
  if (attrStrings) str.add(attrStrings + " ");
  if (kind === "partial") str.add("-");
  str.add("{ ");

  const last = contents.length - 1;
  let lastWasScope = false;
  contents.forEach((elem, i) => {
    if (childScope(elem)) {
      const childScope: Scope = elem;
      const childBlock = scopeToString(childScope, indent + 2);
      !lastWasScope && str.nl();
      str.addBlock(childBlock);
      lastWasScope = true;
    } else {
      lastWasScope && str.add("  ");
      lastWasScope = false;
      const ident: Ident = elem;
      str.add(identShortString(ident));
      if (i < last) str.add(" ");
    }
  });

  if (str.oneLine) {
    str.add(" }");
  } else {
    str.add("}");
  }

  str.add(` #${scope.id}`);

  return str.result;
}

/** name of an identifier, with decls prefixed with '%' */
function identShortString(ident: Ident): string {
  const { kind, originalName } = ident;
  const prefix = kind === "decl" ? "%" : "";
  return `${prefix}${originalName}`;
}

export function identToString(ident?: Ident): string {
  if (!ident) return JSON.stringify(ident);
  const { kind, originalName } = ident;
  const idStr = ident.id ? `#${ident.id}` : "";
  if (kind === "ref") {
    const ref = identToString(ident.refersTo!);
    return `${originalName} ${idStr} -> ${ref}`;
  } else {
    const { mangledName } = ident;
    const mangled = mangledName ? `(${mangledName})` : "";
    return `%${originalName}${mangled} ${idStr} `;
  }
}
