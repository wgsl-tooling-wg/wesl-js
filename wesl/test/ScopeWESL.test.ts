import { findValidRootDecls } from "../BindIdents.ts";
import { scopeToString } from "../debug/ScopeToString.ts";
import type { WeslAST } from "../ParseWESL.ts";
import { resetScopeIds } from "../Scope.ts";
import { parseWESL } from "./TestUtil.ts";
import { assertSnapshot } from "@std/testing/snapshot";

function testParseWESL(src: string): WeslAST {
  resetScopeIds();
  return parseWESL(src);
}

Deno.test("scope from simple fn", async (t) => {
  const src = `
    fn main() {
      var x: i32 = 1;
    }
  `;
  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("scope from fn with reference", async (t) => {
  const src = `
    fn main() {
      var x: i32 = 1;
      x++;
    }
  `;
  const { rootScope } = testParseWESL(src);

  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("two fns", async (t) => {
  const src = `
    fn foo() {}
    fn bar() {}
  `;
  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("two fns, one with a decl", async (t) => {
  const src = `
    fn foo() {
      var a:u32;
    }
    fn bar() {}
  `;
  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("fn ref", async (t) => {
  const src = `
    fn foo() {
      bar();
    }
    fn bar() {}
  `;
  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("struct", async (t) => {
  const src = `
    struct A {
      a: B,
    }
  `;
  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("alias", async (t) => {
  const src = `
    alias A = B;
  `;
  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("switch", async (t) => {
  const src = `
    fn main() {
      var code = 1u;
      switch ( code ) {
        case 5u: { if 1 > 0 { var x = 7;} }
        default: { break; }
      }
    }`;
  const { rootScope } = testParseWESL(src);

  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("for()", async (t) => {
  const src = `
    fn main() {
      var i = 1.0;
      for (var i = 0; i < 10; i++) { }
    }`;
  const { rootScope } = testParseWESL(src);

  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("fn with param", async (t) => {
  const src = `
    fn main(i: i32) {
      var x = 10 + i;
      for (var i = 0; i < x; i++) { }
    }`;
  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("fn decl scope", async (t) => {
  const src = `
    fn main(i: i32) {
      var x = i;
    }`;
  const { rootScope } = testParseWESL(src);
  const decls = findValidRootDecls(rootScope, {});
  const mainIdent = decls[0];
  await assertSnapshot(t, scopeToString(mainIdent.dependentScope!));
});

Deno.test("builtin scope", async (t) => {
  const src = `fn main( @builtin(vertex_index) a: u32) { }`;
  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("builtin enums", async (t) => {
  const src =
    `struct read { a: vec2f } var<storage, read_write> storage_buffer: read;`;
  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("texture_storage_2d", async (t) => {
  const src = `
    @binding(3) @group(0) var tex_out : texture_storage_2d<rgba8unorm, write>;
  `;
  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("ptr 2 params", async (t) => {
  const src = `
    fn foo(ptr: ptr<private, u32>) { }
  `;
  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("ptr 3 params", async (t) => {
  const src = `
    fn foo(ptr: ptr<storage, array<u32, 128>, read>) { }
  `;
  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("larger example", async (t) => {
  const src = `
    struct UBO { width : u32, }

    struct Buffer { weights : array<f32>, }

    @binding(0) @group(0) var<uniform> ubo : UBO;
    @binding(1) @group(0) var<storage, read> buf_in : Buffer;
    @binding(2) @group(0) var<storage, read_write> buf_out : Buffer;
    @binding(3) @group(0) var tex_in : texture_2d<f32>;
    @binding(3) @group(0) var tex_out : texture_storage_2d<rgba8unorm, write>;

    @compute @workgroup_size(64)
    fn import_level(@builtin(global_invocation_id) coord : vec3u) {
      _ = &buf_in;
      let offset = coord.x + coord.y * ubo.width;
      buf_out.weights[offset] = textureLoad(tex_in, vec2i(coord.xy), 0).w;
    }

    @compute @workgroup_size(64)
    fn export_level(@builtin(global_invocation_id) coord : vec3u) {
      if (all(coord.xy < vec2u(textureDimensions(tex_out)))) {
        let dst_offset = coord.x    + coord.y    * ubo.width;
        let src_offset = coord.x*2u + coord.y*2u * ubo.width;

        let a = buf_in.weights[src_offset + 0u];
        let b = buf_in.weights[src_offset + 1u];
        let c = buf_in.weights[src_offset + 0u + ubo.width];
        let d = buf_in.weights[src_offset + 1u + ubo.width];
        let sum = dot(vec4f(a, b, c, d), vec4f(1.0));

        buf_out.weights[dst_offset] = sum / 4.0;

        let probabilities = vec4f(a, a+b, a+b+c, sum) / max(sum, 0.0001);
        textureStore(tex_out, vec2i(coord.xy), probabilities);
      }
    }
  `;

  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("scope with an attribute", async (t) => {
  const src = `
    fn main() {
      @if(foo){ }
    }
  `;
  const { rootScope } = testParseWESL(src);

  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("partial scope", async (t) => {
  const src = `
    fn main() {
      var x = 1;

      @if(false) y = 2;
    }
  `;
  const { rootScope } = testParseWESL(src);

  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("loop scope", async (t) => {
  const src = `
    fn main() {
      let a = 7;
      loop {
        let a = 1;
        continuing {
          let a = 2;
        }
      }
    }
  `;
  const { rootScope } = testParseWESL(src);

  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("nested scope test", async (t) => {
  const src = `
    fn main() {
      let bar = 72;
      if (true) {
        if (true) {
          let new_bar = bar; // Should be 72!
        }
        let bar = 5;
      }
    }
  `;
  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("@if fn", async (t) => {
  const src = `
    const loc = 0;

    @if(true) @fragment
    fn fragmentMain(@location(0) p: vec3f) -> @location(loc) vec4f { 
      let x = p;
    }
  `;
  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("@if const", async (t) => {
  const src = `
    @if(true) const a = 0;
  `;
  const { rootScope } = testParseWESL(src);
  await assertSnapshot(t, scopeToString(rootScope));
});

Deno.test("var<private> a: i32;", async (t) => {
  const src = `
    var<private> a: i32 = 0;
  `;
  const { rootScope } = testParseWESL(src);

  await assertSnapshot(t, scopeToString(rootScope));
});

// Deno.test("", () => {
//   const src = `
//   `;
//   const { rootScope } = testParseWESL(src);
//   expect(scopeToString(rootScope)).toMatchInlineSnapshot('tbd');
// });
