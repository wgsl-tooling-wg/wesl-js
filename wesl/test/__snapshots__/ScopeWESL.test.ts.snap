export const snapshot = {};

snapshot[`scope from simple fn 1`] = `
"{ 
  -{ %main 
    { %x i32 } #2
  } #1
} #0"
`;

snapshot[`scope from fn with reference 1`] = `
"{ 
  -{ %main 
    { %x i32 x } #2
  } #1
} #0"
`;

snapshot[`two fns 1`] = `
"{ 
  -{ %foo 
    {  } #2
  } #1
  -{ %bar 
    {  } #5
  } #4
} #0"
`;

snapshot[`two fns, one with a decl 1`] = `
"{ 
  -{ %foo 
    { %a u32 } #2
  } #1
  -{ %bar 
    {  } #5
  } #4
} #0"
`;

snapshot[`fn ref 1`] = `
"{ 
  -{ %foo 
    { bar } #2
  } #1
  -{ %bar 
    {  } #5
  } #4
} #0"
`;

snapshot[`struct 1`] = `
"{ %A 
  { B } #1
} #0"
`;

snapshot[`alias 1`] = `
"{ %A 
  { B } #1
} #0"
`;

snapshot[`switch 1`] = `
"{ 
  -{ %main 
    { %code code 
      { 
        { %x } #5
      } #4
      {  } #6
    } #2
  } #1
} #0"
`;

snapshot[`for() 1`] = `
"{ 
  -{ %main 
    { %i 
      { %i i i } #4
    } #2
  } #1
} #0"
`;

snapshot[`fn with param 1`] = `
"{ 
  -{ %main 
    { %i i32 %x i 
      { %i i x i } #4
    } #2
  } #1
} #0"
`;

snapshot[`fn decl scope 1`] = `"{ %i i32 %x i } #2"`;

snapshot[`builtin scope 1`] = `
"{ 
  -{ %main 
    { %a u32 } #2
  } #1
} #0"
`;

snapshot[`builtin enums 1`] = `
"{ %read 
  { vec2f } #1
  -{ %storage_buffer 
    { read } #3
  } #2
} #0"
`;

snapshot[`texture_storage_2d 1`] = `
"{ 
  -{ %tex_out 
    { texture_storage_2d rgba8unorm write } #2
  } #1
} #0"
`;

snapshot[`ptr 2 params 1`] = `
"{ 
  -{ %foo 
    { %ptr ptr private u32 } #2
  } #1
} #0"
`;

snapshot[`ptr 3 params 1`] = `
"{ 
  -{ %foo 
    { %ptr ptr storage array u32 read } #2
  } #1
} #0"
`;

snapshot[`larger example 1`] = `
"{ %UBO 
  { u32 } #1
  %Buffer 
  { array f32 } #2
  -{ %ubo 
    { UBO } #4
  } #3
  -{ %buf_in 
    { Buffer } #6
  } #5
  -{ %buf_out 
    { Buffer } #8
  } #7
  -{ %tex_in 
    { texture_2d f32 } #10
  } #9
  -{ %tex_out 
    { texture_storage_2d rgba8unorm write } #12
  } #11
  -{ %import_level 
    { %coord vec3u buf_in %offset coord coord ubo buf_out 
      offset textureLoad tex_in vec2i coord} #14
  } #13
  -{ %export_level 
    { %coord vec3u all coord vec2u textureDimensions tex_out
       
      { %dst_offset coord coord ubo %src_offset coord coord 
        ubo %a buf_in src_offset %b buf_in src_offset %c 
        buf_in src_offset ubo %d buf_in src_offset ubo %sum 
        dot vec4f a b c d vec4f buf_out dst_offset sum 
        %probabilities vec4f a a b a b c sum max sum 
        textureStore tex_out vec2i coord probabilities} #19
    } #17
  } #16
} #0"
`;

snapshot[`scope with an attribute 1`] = `
"{ 
  -{ %main 
    { 
       @if(foo) {  } #4
    } #2
  } #1
} #0"
`;

snapshot[`partial scope 1`] = `
"{ 
  -{ %main 
    { %x 
       @if(false) -{ y } #4
    } #2
  } #1
} #0"
`;

snapshot[`loop scope 1`] = `
"{ 
  -{ %main 
    { %a 
      { %a 
        { %a } #5
      } #4
    } #2
  } #1
} #0"
`;

snapshot[`nested scope test 1`] = `
"{ 
  -{ %main 
    { %bar 
      { 
        { %new_bar bar } #5
        %bar
      } #4
    } #2
  } #1
} #0"
`;

snapshot[`@if fn 1`] = `
"{ 
  -{ %loc 
    {  } #2
  } #1
   @if(true) -{ %fragmentMain loc 
    { %p vec3f vec4f %x p } #4
  } #3
} #0"
`;

snapshot[`@if const 1`] = `
"{ 
   @if(true) -{ %a 
    {  } #2
  } #1
} #0"
`;

snapshot[`var<private> a: i32; 1`] = `
"{ 
  -{ %a 
    { i32 } #2
    {  } #3
  } #1
} #0"
`;
